common/lib/dependabot/clients/bitbucket_server.rb

```diff
@@ -0,0 +1,236 @@
+        api_path += path || commit ? "limit=100" : "?limit=100"
+        response = get(File.join(repo_path, api_path))
+
+        response.body.dig("children", "values")&.map do |file|
+          file["path"] = if path.nil? || path.empty?
+                           file.dig("path", "name")
+                         else
+                           path
+                         end
+          file
+        end
+      end
+
+      # rubocop:enable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+      def fetch_file_contents(_repo_name, commit, path)
+        path = "raw/#{path.gsub(%r{/+$}, '')}?at=#{commit}"
+        response = get(File.join(repo_path, path))
+        response.body
+      end
+
+      def tags(_repo_name)
+        url = File.join(repo_path, "tags?limit=100")
+        response = get(url)
+
+        response.body.fetch("values")
+      end
+
+      def compare(_repo_name, previous_tag, new_tag)
+        path = "compare/commits?to=#{new_tag}&from=#{previous_tag}"
+        response = get(File.join(repo_path, path))
+
+        response.body.fetch("values")
+      end
+
+      def fetch_pull_request_for_branch(branch)
+        path = "/pull-requests?at=refs/heads/#{branch}&direction=outgoing"
+        url = File.join(repo_path, path)
+        response = get(url)
+
+        response.body.dig("values", 0)
+      end
+
+      def create_pull_request(title, from, to, description, reviewers = nil)
+        url = File.join(repo_path, "pull-requests")
+        from_ref = { "id" => from }
+        to_ref   = { "id" => to }
+        body = {
+          title: title,
+          description: description,
+          state: "OPEN",
+          fromRef: from_ref,
+          toRef: to_ref,
+          reviewers: Array(reviewers).map { |r| { user: { name: r } } }
+        }.to_json
+        post(url, body)
+      end
+
+      private
+
+      attr_reader :credentials
+      attr_reader :source
+
+      def request(method, url, body = nil)
+        res = conn.public_send(method) do |req|
+          req.url url
+          req.headers["Content-Type"] = "application/json"
+          req.body = body unless body.nil?
+        end
+
+        handle_response(res)
+      end
+
+      def handle_response(response)
+        raise Unauthorized if response.status == 401
+        raise Forbidden if response.status == 403
+        raise NotFound if response.status == 404
+
+        if response.status >= 400
+          raise "Unhandled Bitbucket error!\n"\
+                "Status: #{response.status}\n"\
+                "Body: #{response.body}"
+        end
+
+        response
+      end
+
+      %i(get post).each do |method|
+        define_method(method.to_s) do |*args|
+          request(method, *args)
+        end
+      end
+
+      def put_multipart(url, body)
+        conn.put do |req|
+          req.url url
+          req.body = body
+          req.headers["Content-Type"] = "multipart/form-data"
+        end
```

Comment:
`handle_response` needs to be used here too, as in `request`, so that an exception is raised if this fails.